crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -46,11 +46,27 @@ impl<'db> UnionBuilder<'db> {
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                self.elements.extend(union.elements(self.db));
+                for element in union.elements(self.db) {
+                    self = self.add(*element);
+                }
             }
             Type::Never => {}
             _ => {
-                self.elements.insert(ty);
+                let mut add = true;
+                let mut remove = vec![];
+                for element in &self.elements {
+                    if ty.is_subtype_of(self.db, *element) {
+                        add = false;
+                    } else if element.is_subtype_of(self.db, ty) {
+                        remove.push(*element);
+                    }
+                }
+                for element in remove {
+                    self.elements.remove(&element);
+                }
+                if add {
+                    self.elements.insert(ty);
+                }
```

Comment:
As soon as you find an element already in the union that you know to be a supertype of the element you're considering adding to the union, you can short-circuit, since you know you won't be adding or removing any elements from the union in that case:

```suggestion
                let mut remove = vec![];
                for element in &self.elements {
                    if ty.is_subtype_of(self.db, *element) {
                        return self;
                    } else if element.is_subtype_of(self.db, ty) {
                        remove.push(*element);
                    }
                }
                for element in remove {
                    self.elements.remove(&element);
                }
                self.elements.insert(ty);
```

Reply:
Ugh, yes thanks! I had this in mind when writing it, then the tests passed and I totally forgot. Will fix.